transactions/src/main/java/org/jboss/as/txn/service/internal/tsr/JCAOrderedLastSynchronizationList.java

```diff
@@ -0,0 +1,142 @@
+                TransactionLogger.ROOT_LOGGER.trace("JCAOrderedLastSynchronizationList.preJcaSyncs.before_completion - Class: " + preJcaSync.getClass() + " HashCode: "
+                    + preJcaSync.hashCode()
+                    + " toString: "
+                    + preJcaSync);
+            }
+            preJcaSync.beforeCompletion();
+        }
+        for (Synchronization jcaSync : jcaSyncs) {
+            if (TransactionLogger.ROOT_LOGGER.isTraceEnabled()) {
+                TransactionLogger.ROOT_LOGGER.trace("JCAOrderedLastSynchronizationList.jcaSyncs.before_completion - Class: " + jcaSync.getClass() + " HashCode: "
+                    + jcaSync.hashCode()
+                    + " toString: "
+                    + jcaSync);
+            }
+            jcaSync.beforeCompletion();
+        }
+    }
+
+    @Override
+    public void afterCompletion(int status) {
+        for (Synchronization preJcaSync : preJcaSyncs) {
+            if (TransactionLogger.ROOT_LOGGER.isTraceEnabled()) {
+                TransactionLogger.ROOT_LOGGER.trace("JCAOrderedLastSynchronizationList.preJcaSyncs.afterCompletion - Class: " + preJcaSync.getClass() + " HashCode: "
+                    + preJcaSync.hashCode()
+                    + " toString: " + preJcaSync);
+            }
+            preJcaSync.afterCompletion(status);
+        }
+        for (Synchronization jcaSync : jcaSyncs) {
+            if (TransactionLogger.ROOT_LOGGER.isTraceEnabled()) {
+                TransactionLogger.ROOT_LOGGER.trace("JCAOrderedLastSynchronizationList.jcaSyncs.afterCompletion - Class: " + jcaSync.getClass() + " HashCode: "
+                    + jcaSync.hashCode()
+                    + " toString: "
+                    + jcaSync);
+            }
+            jcaSync.afterCompletion(status);
+        }
+
+        if (jcaOrderedLastSynchronizations.remove(tx) == null) {
+            // The identifier wasn't stable - scan for it - this can happen in JTS propagation when the UID needs retrieving
+            // from the parent and the parent has been deactivated
+            Transaction altKey = null;
+            Iterator<Entry<Transaction, JCAOrderedLastSynchronizationList>> iterator = jcaOrderedLastSynchronizations.entrySet().iterator();
+            while (altKey == null && iterator.hasNext()) {
+                Map.Entry<Transaction, JCAOrderedLastSynchronizationList> next = iterator.next();
+                if (next.getValue().equals(this)) {
+                    altKey = next.getKey();
+                    iterator.remove();
+                    if (TransactionLogger.ROOT_LOGGER.isTraceEnabled()) {
+                        TransactionLogger.ROOT_LOGGER.tracef("Removed: %s [%s]", System.identityHashCode(tx), tx.toString());
+                    }
+                    break;
+                }
+            }
+
+            if (altKey == null) {
+                TransactionLogger.ROOT_LOGGER.transactionNotFound(tx);
```

Comment:
Jesper gave me a heads up on this actually on the forum (https://developer.jboss.org/message/920250#920250). He correctly reminded me that the identify of the transaction is not guaranteed after completion and hence the .equals() may not be true (its allowed by spec as this is afterCompletion). The algorithm Jesper proposed should never hit that warning but I left it in because, well you never know...